File: 프로그래머스/Level 1/공원 산책.py
def inRange(y, x, r, c):      
    return (y >= 0 and x >= 0 and y<r and x < c)
        

def solution(park, routes):
    cy,cx = -1,-1 # 현재 좌표
    lenY, lenX = len(park), len(park[0])
    
    # 현재 좌표 찾기
    for i in range(lenY):
        for j in range(lenX):
            if park[i][j] == "S":
                cy, cx = i, j
                break
        if cy != -1 and cx != -1:
            break
    
    # 명령을 하나씩 확인하며 진행
    for r in routes:
        d, l = r[0], int(r[2])
        flag = 0
        
        if d == "E":
            if not inRange(cy, cx+l, lenY, lenX):
                    continue
            for i in range(1, l+1):
                if park[cy][cx+i] == "X":
                    flag = 1
                    break
                    
            if flag == 0:
                cx += l
            
        elif d == "W":
            if not inRange(cy, cx-l, lenY, lenX):
                    continue
            for i in range(1, l+1):
                if park[cy][cx-i] == "X":
                    flag = 1
                    break
                    
            if flag == 0:
                cx -= l
            
        elif d == "S":
            if not inRange(cy+l, cx, lenY, lenX):
                    continue
            for i in range(1, l+1):
                if park[cy+i][cx] == "X":
                    flag = 1
                    break
                    
            if flag == 0:
                cy += l
        
        elif d == "N":
            if not inRange(cy-l, cx, lenY, lenX):
                    continue
            for i in range(1, l+1):
                if park[cy-i][cx] == "X":
                    flag = 1
                    break
                    
            if flag == 0:
                cy -= l
                
    return [cy, cx]

File: 프로그래머스/Level 1/공원 산책.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/172928
import Foundation

func solution(_ park:[String], _ routes:[String]) -> [Int] {
    var answer:[Int] = [-1, -1]
    var board:[[String]] = []
    let row = park.count
    let col = park[0].count

    func getBoard() {
        for i in 0..<row {
            board.append(park[i].map{ String($0)})
        }
    }
    
    func getCurPosition() -> [Int] {
        var tmp:[Int] = [-1, -1]
        
        for i in 0..<row {
            for j in 0..<col {
                if board[i][j] == "S" {
                    tmp[0] = i
                    tmp[1] = j
                    break
                }
            }
        }
        
        return tmp
    }
    
    func inRange(_ x: Int, _ y: Int) -> Bool {
        return x >= 0 && y >= 0 && x < row && y < col
    }
    
    getBoard()
    answer = getCurPosition()
    
    for route in routes {
        let dir = String(route.first!)
        let n = Int(String(route.last!))!
        
        if dir == "E" {
            if !inRange(answer[0], answer[1]+n) {
                continue
            }
            var flag = 0
            for d in 1...n {
                if board[answer[0]][answer[1]+d] == "X" {
                    flag = 1
                    break
                }
            }
            
            if flag == 0 {
                answer[1] += n
            }
        } else if dir == "W" {
            if !inRange(answer[0], answer[1]-n) {
                continue
            }
            var flag = 0
            for d in 1...n {
                if board[answer[0]][answer[1]-d] == "X" {
                    flag = 1
                    break
                }
            }
            
            if flag == 0 {
                answer[1] -= n
            }
        } else if dir == "S" {
            if !inRange(answer[0]+n, answer[1]) {
                continue
            }
            var flag = 0
            for d in 1...n {
                if board[answer[0]+d][answer[1]] == "X" {
                    flag = 1
                    break
                }
            }
            
            if flag == 0 {
                answer[0] += n
            }
        } else if dir == "N" {
            if !inRange(answer[0]-n, answer[1]) {
                continue
            }
            var flag = 0
            for d in 1...n {
                if board[answer[0]-d][answer[1]] == "X" {
                    flag = 1
                    break
                }
            }
            
            if flag == 0 {
                answer[0] -= n
            }
        }
    }
    return answer
}